src/psyclone/domain/nemo/transformations/nemo_arrayaccess2loop_trans.py

```diff
@@ -0,0 +1,328 @@
+                nloops += 1
+            location = location.parent.parent
+
+        # Look up the loop variable in the symbol table. If it does
+        # not exist then create it.
+        try:
+            loop_variable_symbol = symbol_table.lookup(loop_variable_name)
+        except KeyError:
+            # Add loop variable as it does not already exist
+            loop_variable_symbol = DataSymbol(loop_variable_name, INTEGER_TYPE)
+            symbol_table.add(loop_variable_symbol)
+
+        # Replace array access loop variable.
+        for array in assignment.walk(ArrayReference):
+            if not array.ancestor(ArrayReference):
+                # This is not a nested access e.g. a(b(n)).
+                array.children[array_index] = Reference(loop_variable_symbol)
+
+        # As this transformation affects ancestor nodes of the
+        # supplied node we need to be careful here not to lose
+        # existing ancestor nodes as they might be
+        # referenced. Therefore use detach() and insert,() rather than
+        # copy() and replace_with().
+
+        # Find location.parent and location_index
+        loc_parent = location.parent
+        loc_index = location.position
+
+        # Create the new single-trip loop and add its children.
+        step = Literal("1", INTEGER_TYPE)
+        loop = NemoLoop.create(loop_variable_symbol, node_copy,
+                               node_copy.copy(), step, [location.detach()])
+        # Replace the original assignment with a loop containing the
+        # modified assignment.
+        loc_parent.children.insert(loc_index, loop)
+
+        if nloops == 0 and not assignment.walk(Range):
+            # There were previously no loops therefore there was not
+            # inlined kernel node and there are no range nodes in the
+            # array reference, so we now need to add an inlined
+            # kernel.
+            CreateNemoKernelTrans().apply(assignment.parent)
+
+    def validate(self, node, options=None):
+        '''Perform various checks to ensure that it is valid to apply the
+        NemoArrayAccess2LoopTrans transformation to the supplied PSyIR Node.
+
+        :param node: the node that is being checked.
+        :type node: :py:class:`psyclone.psyir.nodes.Node`
+        :param options: a dictionary with options for \
+            transformations. No options are used in this \
+            transformation. This is an optional argument that defaults \
+            to None.
+        :type options: dict of string:values or None
+
+        '''
+        # Not a PSyIR node
+        if not isinstance(node, Node):
+            raise TransformationError(
+                "Error in NemoArrayAccess2LoopTrans transformation. The "
+                "supplied node argument should be a PSyIR Node, but found "
+                "'{0}'.".format(type(node).__name__))
+        # Not within an array reference
+        if not node.parent or not isinstance(node.parent, ArrayReference):
```

Comment:
s/ArrayReference/ArrayMixin/ perhaps so as to allow for array members of derived types?

Reply:
I decided to stick with plain arrays for the moment as there is no need to support them at the moment and have added a test to make this clear.

Reply:
Agreed but since we're here, please could you change this to ArrayMixin anyway?